server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -669,56 +683,177 @@ public String toString() {
+        private final Iterator<StoreFileMetaData> remainingFiles;
+        private StoreFileMetaData currentFile;
+        private InputStreamIndexInput currentInput = null;
+        private long currentChunkPosition = 0;
+        private final Deque<byte[]> recycledBuffers = ConcurrentCollections.newDeque();
+        private final AtomicReference<Tuple<StoreFileMetaData, Exception>> error = new AtomicReference<>();
+
+        MultiFileSender(Store store, IntSupplier translogOps, StoreFileMetaData[] files, ActionListener<Void> listener) {
+            super(ActionListener.notifyOnce(listener));
+            this.store = store;
+            this.translogOps = translogOps;
+            this.remainingFiles = Arrays.asList(files).iterator();
+        }
+
+        @Override
+        protected void doRun() throws Exception {
+            assert ThreadPool.assertCurrentMethodIsNotCalledRecursively();
+            assert Transports.assertNotTransportThread(RecoverySourceHandler.this + "[send file chunk]");
+            while (true) {
+                assert semaphore.availablePermits() == 0;
+                if (error.get() != null) {
+                    handleErrorOnSendFiles(store, error.get().v2(), new StoreFileMetaData[]{error.get().v1()});
+                    throw error.get().v2();
+                }
+                final FileChunk chunk = readNextChunk();
+                if (chunk == null) {
+                    semaphore.release(); // allow other threads respond if we are not done yet.
+                    if (requestSeqIdTracker.getMaxSeqNo() == requestSeqIdTracker.getProcessedCheckpoint() && semaphore.tryAcquire()) {
+                        listener.onResponse(null);
                     }
-                    final long requestFilePosition = position;
-                    cancellableThreads.executeIO(() ->
-                        recoveryTarget.writeFileChunk(md, requestFilePosition, content, lastChunk, translogOps.getAsInt(),
-                            ActionListener.wrap(
-                                r -> requestSeqIdTracker.markSeqNoAsProcessed(requestSeqId),
-                                e -> {
-                                    error.compareAndSet(null, Tuple.tuple(md, e));
-                                    requestSeqIdTracker.markSeqNoAsProcessed(requestSeqId);
+                    break;
+                }
+                final long requestSeqId = requestSeqIdTracker.generateSeqNo();
+                cancellableThreads.execute(() ->
+                    recoveryTarget.writeFileChunk(chunk.md, chunk.position, chunk.content, chunk.lastChunk, translogOps.getAsInt(),
+                        ActionListener.wrap(
+                            r -> {
+                                chunk.close(); // release the buffer so we can reuse to reduce allocation
+                                requestSeqIdTracker.markSeqNoAsProcessed(requestSeqId);
+                                if (canSendMore() && semaphore.tryAcquire()) {
```

Comment:
The way we check `canSendMore()` here could lead to spawning more requests than the max (at least 1). Imagine following sequence:

1. Transport: Read `getMaxSeqNo()` here.
2. Generic: A new request is made, pulling a new number, incrementing maxSeqNo.
3. Generic: The semaphore is released.
4. Transport: Read `getProcessedCheckpoint` and spawn `sendFileExecutor.execute(this).

We then risk estimating number of active requests to be smaller than what it really is and re-execute `this`.

I believe removing the `canSendMore()` check here and doing that at the top of the while (true) loop instead will fix this (since then maxSeqNo cannot increase).

Reply:
Great catch.